### DISCUSSION TITLE
Max call stack size exceeded from `makeTimingReport`

### DISCUSSION THREAD

**Discussion by simonhaenisch** [renovatebot/renovate#29450] [Request Help]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Bitbucket Server

### Please tell us more about your question or problem

The problem is basically this `Math.max` call in `makeTimingReport`:

https://github.com/renovatebot/renovate/blob/d8c344012d9c17103ffbf88eead9f868abce5d4d/lib/util/stats.ts#L19

When `data` has too many items which was the case for one of our repos, this causes `Math.max` to receive a very large amount of arguments, which apparently can end in the "maximum call stack size exceeded" error.

I found some info about it on [this StackOverflow question](https://stackoverflow.com/questions/42623071/maximum-call-stack-size-exceeded-with-math-min-and-math-max):

> Not really. These functions are not recursive (see v8 sources in the comment above). The problem is in the amount of function parameters after destructuring.

We were able to prevent the error by patching `node_modules/renovate/dist/util/stats.js` by replacing `Math.max` with a custom function that would take a reference to the array as an argument instead of each array element as its own argument, and loop over the array.

**Comment by simonhaenisch** [comment]
Happy to make a PR for this btw, but I'd like some feedback first whether it makes sense to add a custom `max` function, or if it's better to use some package like [lodash.max](https://www.npmjs.com/package/lodash.max)?

**Reply by secustor** [reply]
We prefer to not introduce new dependencies if it's easy avoidable. So if you use packages we already import use that, else please create a custom one.
